File: dotnet/ThirtyFour.UI/Window/TitleMatcher.cs
using System;
using System.Collections.Generic;
using OpenQA.Selenium;
using ThirtyFour.UI.Util;

namespace ThirtyFour.UI.Window
{
    /// <summary>
    /// The TitleMatcher class matches and switches to the first popup window it finds which matches the provided title
    /// </summary>
    public class TitleMatcher : IMatcher
    {
        private string title;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="title">The title of the window to match</param>
        public TitleMatcher(string title)
        {
            this.title = title;
        }

        /// <summary>
        /// Switches to the first popup window found which matches the provided title
        /// </summary>
        /// <param name="driver">The IWebDriver instance</param>
        /// <param name="timeoutInSeconds">The number of seconds to keep trying to find the matching popup window</param>
        public void MatchWindow(IWebDriver driver, double timeoutInSeconds)
        {

            string currentWindowHandle = driver.CurrentWindowHandle;

            bool found = Utils.RetryUntilTimeout(TimeSpan.FromSeconds(timeoutInSeconds), TimeSpan.FromMilliseconds(100), () => {
                IReadOnlyCollection<string> windows = driver.WindowHandles;

                foreach (String window in windows)
                {
                    driver.SwitchTo().Window(window);
                    if (driver.Title.Equals(this.title))
                    {
                        return true;
                    }
                    driver.SwitchTo().Window(currentWindowHandle);
                }
                return false;
            });

            if (!found)
            {
                throw new Exception("Unable to find Window");
            }
        }
    }
}

File: java/src/main/java/com/github/danhumphrey/thirtyfour/ui/window/TitleMatcher.java
package com.github.danhumphrey.thirtyfour.ui.window;

import java.util.Set;
import java.util.concurrent.Callable;

import org.openqa.selenium.WebDriver;

import com.github.danhumphrey.thirtyfour.ui.util.Utils;

/**
 * The TitleMatcher matches the first browser window it finds with a matching title
 *
 */
public class TitleMatcher implements Matcher {
	
	private String title;
	
	/**
	 * Constructor
	 * @param title the title of the window you want to match
	 */
	public TitleMatcher(String title) {
		this.title = title;
	}
	
	public class MatcherTask implements Callable<Boolean> {
	    
		private WebDriver driver;
		
		public MatcherTask(WebDriver driver) {
			this.driver = driver;
		}
		
		public Boolean call() throws InterruptedException {
			
			String currentHandle = driver.getWindowHandle();
			
	    	Set<String> handles = driver.getWindowHandles();
	    	for(String handle : handles) {
	    		
	    		driver.switchTo().window(handle);
	    		if(driver.getTitle().equals(title)) {
	    			return true;
	    		}
	    		driver.switchTo().window(currentHandle);
	    	}
	    	return false;
	    }
	}
	
	@Override
	public void matchWindow(WebDriver driver, long timeoutInMilliseconds) throws Exception {
		
		boolean found = Utils.retryUntilTimeout(timeoutInMilliseconds, 100, new MatcherTask(driver));
		
		if (!found) {
            throw new Exception("Unable to find a matching window");
        }
			
	}

}